File: skills/tduck/scripts/replace_endpoint.js
#!/usr/bin/env node

/**
 * ==============================================================================
 * TDuck Skill 服务端点一键批量替换脚本 (Node.js 原生版)
 * 
 * 纯原生 Node.js 实现，零外部依赖，跨平台兼容 macOS、Linux 与 Windows。
 * 用于将本 Skill 仓库中所有文件内的 TDuck 服务地址 / MCP 端点批量替换为您自己的实际部署地址。
 *
 * 用法示例：
 *   1. 命令行直接指定新地址：
 *      node replace_endpoint.js https://x.tduckcloud.com/tduck-api
 *    
 *
 *   2. 带参数指定：
 *      node replace_endpoint.js --url https://x.tduckcloud.com/tduck-api
 *
 *   3. 预览模式（仅检查匹配文件，不实际写入）：
 *      node replace_endpoint.js --url https://x.tduckcloud.com/tduck-api --dry-run
 *
 *   4. 交互式运行：
 *      node replace_endpoint.js
 * ==============================================================================
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');

// 需扫描的文件后缀
const TARGET_EXTENSIONS = new Set([
  '.md', '.json', '.py', '.sh', '.js', '.ts', '.yml', '.yaml', '.txt', '.svg'
]);

// 需忽略的目录
const IGNORE_DIRS = new Set([
  '.git', '.idea', '.vscode', 'node_modules', 'target', 'dist', 'build', '__pycache__', '.gemini'
]);

// 排除自身脚本文件名
const IGNORE_FILES = new Set([
  'replace_endpoint.js',
  'replace_endpoint.sh',
  'replace_endpoint.py'
]);

// 常见的预置旧服务端点地址
const KNOWN_DEFAULT_MCP_URLS = [
  'https://dev.tduckcloud.com/tduck-api/mcp',
  'http://localhost:8996/tduck-api/mcp'
];

/**
 * 规范化输入的 URL，拆解出 base_url 与 mcp_url
 */
function normalizeUrls(inputUrl) {
  if (!inputUrl) return { baseUrl: '', mcpUrl: '' };
  let cleaned = inputUrl.trim().replace(/\/+$/, '');
  let baseUrl = '';
  let mcpUrl = '';

  if (cleaned.endsWith('/mcp')) {
    mcpUrl = cleaned;
    baseUrl = cleaned.slice(0, -4);
  } else {
    baseUrl = cleaned;
    mcpUrl = `${cleaned}/mcp`;
  }
  return { baseUrl, mcpUrl };
}

/**
 * 向上自动定位最顶层 tduckx-skills 仓库根目录
 */
function findRepoRoot() {
  let current = path.resolve(__dirname);
  let topRoot = current;

  while (current !== path.dirname(current)) {
    const hasGit = fs.existsSync(path.join(current, '.git'));
    const hasSkillsAndReadme = fs.existsSync(path.join(current, 'skills')) && fs.existsSync(path.join(current, 'README.md'));
    const hasSkillMd = fs.existsSync(path.join(current, 'SKILL.md'));

    if (hasGit || hasSkillsAndReadme || (hasSkillMd && fs.existsSync(path.join(current, 'skills')))) {
      topRoot = current;
    }
    current = path.dirname(current);
  }
  return topRoot;
}

/**
 * 从现有配置文件和内容中收集已存在的所有旧 MCP 地址
 */
function collectOldUrls(rootDir) {
  const urlSet = new Set(KNOWN_DEFAULT_MCP_URLS);

  const candidateFiles = [
    path.join(rootDir, 'trae/tduck/.mcp.json'),
    path.join(rootDir, 'workbuddy/connector/tduck/mcp.json'),
    path.join(rootDir, '.cursor/mcp.json'),
    path.join(rootDir, 'skills/tduck/SKILL.md'),
    path.join(rootDir, 'SKILL.md'),
    path.join(rootDir, 'README.md')
  ];

  for (const filePath of candidateFiles) {
    if (fs.existsSync(filePath)) {
      try {
        const content = fs.readFileSync(filePath, 'utf-8');
        const matches = content.matchAll(/https?:\/\/[^\s"'`)]+\/tduck-api(\/mcp)?/g);
        for (const m of matches) {
          const matchedUrl = m[0];
          const normalized = normalizeUrls(matchedUrl);
          if (normalized.mcpUrl) {
            urlSet.add(normalized.mcpUrl);
          }
        }
      } catch {
        // 忽略读取错误
      }
    }
  }

  return Array.from(urlSet);
}

/**
 * 递归收集所有目标文件
 */
function collectFiles(dir, fileList = []) {
  let entries = [];
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return fileList;
  }

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (!IGNORE_DIRS.has(entry.name)) {
        collectFiles(fullPath, fileList);
      }
    } else if (entry.isFile()) {
      if (IGNORE_FILES.has(entry.name)) {
        continue;
      }
      const ext = path.extname(entry.name).toLowerCase();
      if (TARGET_EXTENSIONS.has(ext)) {
        fileList.push(fullPath);
      }
    }
  }
  return fileList;
}

/**
 * 对单文件执行替换
 */
function replaceInFile(filePath, pairs, dryRun = false) {
  let content = '';
  try {
    content = fs.readFileSync(filePath, 'utf-8');
  } catch {
    return 0;
  }

  let totalMatches = 0;
  let newContent = content;

  for (const [oldStr, newStr] of pairs) {
    if (!oldStr || oldStr === newStr) continue;
    if (newContent.includes(oldStr)) {
      const parts = newContent.split(oldStr);
      const count = parts.length - 1;
      totalMatches += count;
      newContent = parts.join(newStr);
    }
  }

  if (totalMatches > 0 && !dryRun) {
    fs.writeFileSync(filePath, newContent, 'utf-8');
  }

  return totalMatches;
}

/**
 * 执行批量替换流程
 */
function runBatchReplace({ newUrlInput, customRootDir, dryRun = false }) {
  const rootDir = customRootDir ? path.resolve(customRootDir) : findRepoRoot();
  const newUrls = normalizeUrls(newUrlInput);
  const oldMcpUrls = collectOldUrls(rootDir);

  console.log('='.repeat(65));
  console.log('  TDuck Skill 服务端点批量替换工具 (Node.js 原生版)');
  console.log('='.repeat(65));
  console.log(`目标根目录 : ${rootDir}`);
  console.log(`运行模式   : ${dryRun ? '[DRY RUN 预览模式]' : '[写入模式]'}`);
  console.log('─'.repeat(65));
  console.log(`更新目标 MCP : ${newUrls.mcpUrl}`);
  console.log(`更新目标 Base: ${newUrls.baseUrl}`);
  console.log('─'.repeat(65));

  // 构建替换对：先替换具体的 /mcp 地址，再替换 Base 地址
  const replacePairs = [];
  for (const oldMcp of oldMcpUrls) {
    const oldNorm = normalizeUrls(oldMcp);
    if (oldNorm.mcpUrl && oldNorm.mcpUrl !== newUrls.mcpUrl) {
      replacePairs.push([oldNorm.mcpUrl, newUrls.mcpUrl]);
    }
    if (oldNorm.baseUrl && oldNorm.baseUrl !== newUrls.baseUrl) {
      replacePairs.push([oldNorm.baseUrl, newUrls.baseUrl]);
    }
  }

  // 去重并按长度降序排列（确保先匹配长字符串）
  const uniquePairs = [];
  const seenOld = new Set();
  replacePairs.sort((a, b) => b[0].length - a[0].length);
  for (const [oldStr, newStr] of replacePairs) {
    if (!seenOld.has(oldStr)) {
      seenOld.add(oldStr);
      uniquePairs.push([oldStr, newStr]);
    }
  }

  if (uniquePairs.length === 0) {
    console.log('提示: 当前仓库中所有配置地址与目标新地址一致，无需替换。');
    console.log('='.repeat(65));
    return;
  }

  const files = collectFiles(rootDir);
  let totalFilesChanged = 0;
  let totalReplacements = 0;

  for (const filePath of files) {
    const count = replaceInFile(filePath, uniquePairs, dryRun);
    if (count > 0) {
      const relPath = path.relative(rootDir, filePath);
      console.log(` ✓ [${count} 处] ${relPath}`);
      totalFilesChanged += 1;
      totalReplacements += count;
    }
  }

  console.log('─'.repeat(65));
  if (dryRun) {
    console.log(`预览完成！共发现 ${totalFilesChanged} 个文件待更新，合计 ${totalReplacements} 处地址。`);
    console.log('移除 --dry-run 参数后即可实际应用修改。');
  } else {
    console.log(`替换成功！已完成 ${totalFilesChanged} 个文件的更新，共替换 ${totalReplacements} 处端点地址。`);
  }
  console.log('='.repeat(65));
}

/**
 * 命令行参数解析与交互入口
 */
async function main() {
  const args = process.argv.slice(2);
  let newUrl = null;
  let dryRun = false;
  let customRootDir = null;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--dry-run') {
      dryRun = true;
    } else if (arg === '--url' && i + 1 < args.length) {
      newUrl = args[++i];
    } else if (arg.startsWith('--url=')) {
      newUrl = arg.slice(6);
    } else if (arg === '--dir' && i + 1 < args.length) {
      customRootDir = args[++i];
    } else if (arg.startsWith('--dir=')) {
      customRootDir = arg.slice(6);
    } else if (arg === '-h' || arg === '--help') {
      console.log(`
TDuck Skill 服务端点批量替换工具 (Node.js 原生版)

用法:
  node replace_endpoint.js [URL] [选项]

选项:
  --url <url>      新的 TDuck 服务地址（如 https://form.example.com/tduck-api）
  --dry-run        预览模式，只查看将要修改的文件，不实际写入
  --dir <path>     指定目标根目录（默认自动查找仓库根目录）
  -h, --help       显示帮助信息

示例:
  node replace_endpoint.js https://form.example.com/tduck-api
  node replace_endpoint.js --url https://form.example.com/tduck-api --dry-run
`);
      process.exit(0);
    } else if (!arg.startsWith('-') && !newUrl) {
      newUrl = arg;
    }
  }

  if (!newUrl) {
    const rootDir = customRootDir ? path.resolve(customRootDir) : findRepoRoot();
    const oldMcpUrls = collectOldUrls(rootDir);

    console.log('='.repeat(65));
    console.log('  TDuck Skill 服务端点批量替换工具 (Node.js 原生版)');
    console.log('='.repeat(65));
    console.log(`当前检测到的旧端点地址:\n  ${oldMcpUrls.join('\n  ')}\n`);

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });

    const question = (query) => new Promise((resolve) => rl.question(query, resolve));
    try {
      const input = await question('请输入新的 TDuck 服务地址 (如 https://form.example.com/tduck-api): ');
      newUrl = input.trim();
    } finally {
      rl.close();
    }
  }

  if (!newUrl) {
    console.error('错误: 未输入有效的新地址！');
    process.exit(1);
  }

  runBatchReplace({ newUrlInput: newUrl, customRootDir, dryRun });
}

main().catch((err) => {
  console.error('执行失败:', err);
  process.exit(1);
});

File: workbuddy/skill/tduck-form/scripts/replace_endpoint.py
#!/usr/bin/env python3
"""
TDuck Skill 服务端点一键批量替换脚本

用于将本 Skill 仓库中所有文件内的 TDuck 服务地址 / MCP 端点批量替换为您自己的实际部署地址。

默认原地址：
    MCP 端点: http://localhost:8996/tduck-api/mcp
    API 根路径: http://localhost:8996/tduck-api

用法示例：
    # 1. 命令行直接指定新 MCP 地址：
    python3 replace_endpoint.py https://form.example.com/tduck-api/mcp

    # 2. 指定参数：
    python3 replace_endpoint.py --url https://form.example.com/tduck-api/mcp

    # 3. 试运行模式（仅查看将要修改的文件，不实际写入）：
    python3 replace_endpoint.py --url https://form.example.com/tduck-api/mcp --dry-run

    # 4. 交互式运行：
    python3 replace_endpoint.py
"""

from __future__ import annotations

import argparse
import os
import sys

# 默认扫描的文件后缀
TARGET_EXTENSIONS = {".md", ".json", ".py", ".yml", ".yaml", ".txt", ".svg"}

# 需忽略的目录
IGNORE_DIRS = {".git", ".idea", ".vscode", "__pycache__", "node_modules", "target"}

# 默认原始地址
OLD_MCP_URL = "http://localhost:8996/tduck-api/mcp"
OLD_BASE_URL = "http://localhost:8996/tduck-api"


def normalize_urls(input_url: str) -> tuple[str, str]:
    """根据输入的 URL 规范化出 base_url 与 mcp_url。"""
    cleaned = input_url.strip().rstrip("/")
    if cleaned.endswith("/mcp"):
        mcp_url = cleaned
        base_url = cleaned[:-4]
    else:
        base_url = cleaned
        mcp_url = f"{cleaned}/mcp"
    return base_url, mcp_url


def find_repo_root() -> str:
    """自动定位 tduckx-skill 根目录。"""
    current = os.path.abspath(os.path.dirname(__file__))
    while current != os.path.dirname(current):
        if os.path.exists(os.path.join(current, "SKILL.md")) and (
            os.path.exists(os.path.join(current, "skills")) or os.path.exists(os.path.join(current, "trae"))
        ):
            return current
        current = os.path.dirname(current)
    # 如果找不到则以当前脚本所在目录为准
    return os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))


def replace_in_file(file_path: str, old_new_pairs: list[tuple[str, str]], dry_run: bool = False) -> int:
    """在指定文件中执行字符串替换。"""
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
    except (UnicodeDecodeError, IOError):
        return 0

    original_content = content
    total_matches = 0

    for old_str, new_str in old_new_pairs:
        if old_str in content:
            matches = content.count(old_str)
            total_matches += matches
            content = content.replace(old_str, new_str)

    if total_matches > 0 and not dry_run:
        with open(file_path, "w", encoding="utf-8") as f:
            f.write(content)

    return total_matches


def run_batch_replace(new_url_input: str, root_dir: str | None = None, dry_run: bool = False) -> None:
    if not root_dir:
        root_dir = find_repo_root()

    new_base_url, new_mcp_url = normalize_urls(new_url_input)

    print("=" * 65)
    print("  TDuck Skill 服务端点批量替换工具")
    print("=" * 65)
    print(f"目标根目录 : {root_dir}")
    print(f"模式       : {'[DRY RUN 预览模式]' if dry_run else '[写入模式]'}")
    print("─" * 65)
    print(f"旧 MCP 地址 : {OLD_MCP_URL}  -->  {new_mcp_url}")
    print(f"旧 Base 地址: {OLD_BASE_URL}  -->  {new_base_url}")
    print("─" * 65)

    # 替换顺序：先替换更长更精确的 /mcp，再替换 base_url
    pairs = [
        (OLD_MCP_URL, new_mcp_url),
        (OLD_BASE_URL, new_base_url),
    ]

    total_files_changed = 0
    total_replacements = 0

    for root, dirs, files in os.walk(root_dir):
        # 过滤忽略目录
        dirs[:] = [d for d in dirs if d not in IGNORE_DIRS]

        for file in files:
            # 排除替换脚本自身，避免破坏脚本内的默认常量
            if file == "replace_endpoint.py":
                continue

            ext = os.path.splitext(file)[1].lower()
            if ext in TARGET_EXTENSIONS:
                file_path = os.path.join(root, file)
                count = replace_in_file(file_path, pairs, dry_run=dry_run)
                if count > 0:
                    rel_path = os.path.relpath(file_path, root_dir)
                    print(f" ✓ [{count} 处] {rel_path}")
                    total_files_changed += 1
                    total_replacements += count

    print("─" * 65)
    if dry_run:
        print(f"预览完成！共发现 {total_files_changed} 个文件需要修改，合计 {total_replacements} 处地址。")
        print("去掉 --dry-run 参数即可立即应用修改。")
    else:
        print(f"替换成功！已成功修改 {total_files_changed} 个文件，共完成 {total_replacements} 处地址替换。")
    print("=" * 65)


def main() -> None:
    parser = argparse.ArgumentParser(
        description="一键批量替换 TDuck Skill 中的 MCP / 服务地址"
    )
    parser.add_argument(
        "url",
        nargs="?",
        default=None,
        help="新的 TDuck 服务地址（如 https://form.example.com/tduck-api/mcp 或 https://form.example.com/tduck-api）",
    )
    parser.add_argument(
        "--url",
        dest="flag_url",
        help="新的 TDuck 服务地址",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="预览模式，仅打印匹配的文件而不实际修改",
    )
    parser.add_argument(
        "--dir",
        dest="target_dir",
        default=None,
        help="指定执行替换的目标目录（默认自动查找 tduckx-skill 根目录）",
    )

    args = parser.parse_args()
    target_url = args.flag_url or args.url

    if not target_url:
        print("=" * 65)
        print("  TDuck Skill 服务端点批量替换工具")
        print("=" * 65)
        print(f"当前默认 MCP 地址: {OLD_MCP_URL}")
        print()
        try:
            target_url = input("请输入您的新 TDuck 服务地址 (如 https://form.example.com/tduck-api): ").strip()
        except (KeyboardInterrupt, EOFError):
            print("\n已取消操作。")
            sys.exit(0)

    if not target_url:
        print("错误: 未提供有效的新地址！")
        sys.exit(1)

    run_batch_replace(target_url, root_dir=args.target_dir, dry_run=args.dry_run)


if __name__ == "__main__":
    main()
